libs/pbd/fpu.cc

```diff
@@ -49,35 +45,32 @@ using namespace std;
 FPU* FPU::_instance (0);
 
 #if ( (defined __x86_64__) || (defined __i386__) || (defined _M_X64) || (defined _M_IX86) ) // ARCH_X86
-#ifndef PLATFORM_WINDOWS
 
-/* use __cpuid() as the name to match the MSVC/mingw intrinsic */
-
-static void
-__cpuid(int regs[4], int cpuid_leaf)
+#ifdef PLATFORM_WINDOWS
+/* Use MSVC/mingw intrinsic to get CPUID */
+#include <intrin.h>
+static void cpuid(int reg[4], int cpuid_leaf)
 {
-	asm volatile (
-#if defined(__i386__)
-			"pushl %%ebx;\n\t"
-#endif
-			"cpuid;\n\t"
-			"movl %%eax, (%1);\n\t"
-			"movl %%ebx, 4(%1);\n\t"
-			"movl %%ecx, 8(%1);\n\t"
-			"movl %%edx, 12(%1);\n\t"
-#if defined(__i386__)
-			"popl %%ebx;\n\t"
-#endif
-			:"=a" (cpuid_leaf) /* %eax clobbered by CPUID */
-			:"S" (regs), "a" (cpuid_leaf)
-			:
-#if !defined(__i386__)
-			"%ebx",
-#endif
-			"%ecx", "%edx", "memory");
+	__cpuid(reg, cpuid_leaf);
 }
-
-#endif /* !PLATFORM_WINDOWS */
+/* Get extended features of CPUID */
+static void cpuidex(int reg[4], int cpuid_leaf, int cpuid_level)
+{
+	__cpuidex(reg, cpuid_leaf, cpuid_level);
+}
+#else
+#include <cpuid.h>
```

Comment:
Do we need  a fallback to the old asm code in case `cpuid.h` is n/a?

Reply:
Possible. However, I have tested this with gcc 6.3.0, which worked. Looking at commit logs, `cpuid.h` goes back to the year [2007](https://github.com/gcc-mirror/gcc/commit/b3172cabd267a6e6dd535c565cb7be297907f9e8) on GCC, and [2011](https://github.com/microsoft/clang/commit/b816276fbbf719342540f4db696a908e2fabd6e0) for clang.  I can leave the old inline assembly, but that will require us to maintain it over time. So, I am leaning against having a fallback version at this moment.

Reply:
indeed new enough.